sources/__tests__/validation.test.ts

```diff
@@ -14,501 +18,761 @@ describe("validation", () => {
+          },
+          list: {
+            stable: { react: "npm:18.0.0" },
+          },
+        });
+
+        await workspace.yarn.catalogs.apply();
+
+        await workspace.writeJson("package.json", {
+          name: "test-package",
+          version: "1.0.0",
+          private: true,
+          dependencies: { react: "17.0.0" },
+        });
+
+        const { stderr } = await workspace.yarn.add("react@17.0.0");
+        expect(stderr).toBe("");
+      });
     });
 
-    it("should calculate validation level only within default groups", async () => {
-      workspace = await createTestWorkspace();
+    describe("workspace pattern matching", () => {
+      it("should skip validation when no pattern matches", async () => {
+        workspace = await createTestWorkspace();
 
-      await workspace.writeCatalogsYml({
-        options: {
-          default: ["beta", "experimental"],
-          validation: {
-            stable: "strict",
-            beta: "off",
-            experimental: "off",
+        await workspace.writeCatalogsYml({
+          options: {
+            default: ["stable"],
           },
-        },
-        list: {
-          stable: {
-            react: "npm:18.0.0",
+          validation: [
+            {
+              workspaces: ["packages-*"],
+              rules: { catalog_protocol_usage: "strict" },
+            },
+          ],
+          list: {
+            stable: { react: "npm:18.0.0" },
           },
-          beta: {
-            react: "npm:18.2.0",
+        });
+
+        await workspace.yarn.catalogs.apply();
+
+        await workspace.writeJson("package.json", {
+          name: "test-package",
+          version: "1.0.0",
+          private: true,
+          dependencies: { react: "17.0.0" },
+        });
+
+        const { stdout } = await workspace.yarn.install();
+        expect(stdout).not.toContain(validationMessage("react"));
+      });
+
+      it("should apply validation when pattern matches with wildcard", async () => {
+        workspace = await createTestWorkspace();
+
+        await workspace.writeCatalogsYml({
+          options: {
+            default: ["stable"],
           },
-          experimental: {
-            react: "npm:19.0.0-alpha",
+          validation: [
+            {
+              workspaces: ["test-*"],
+              rules: { catalog_protocol_usage: "strict" },
+            },
+          ],
+          list: {
+            stable: { react: "npm:18.0.0" },
           },
-        },
+        });
+
+        await workspace.yarn.catalogs.apply();
+
+        await workspace.writeJson("package.json", {
+          name: "test-package",
+          version: "1.0.0",
+          private: true,
+          dependencies: { react: "17.0.0" },
+        });
+
+        await expect(workspace.yarn.install()).rejects.toThrow();
       });
 
-      await workspace.yarn.catalogs.apply();
+      it("should use the first matching rule when multiple rules could match", async () => {
```

Comment:
Could you add more test cases for the “first matching rule wins” behavior? For example, we could test cases where a `*` appears before a more specific pattern, and cases where two rules conflict for the same pattern.

Reply:
👍 I have added additional test cases

- "should apply broader rule when it appears before a more specific pattern"
- "should use the first rule when two conflicting rules have the same pattern
- "should use the first rule when conflicting rules have overlapping wildcard patterns"

https://github.com/toss/yarn-plugin-catalogs/pull/40/commits/1d6d7b49b7abd2b65db841356bb0c73e90e73a7d